File: backend/app/services/system_design_sections.py
# PrepBench - Copyright (c) 2026 Nimish Kanungo
# Licensed under the PolyForm Noncommercial License 1.0.0 (see LICENSE).
# Commercial use requires a separate licence from the copyright holder.

"""
The parts of a written system-design answer, in the order an interviewer expects them.

The plan's chain is prompt -> requirements -> architecture -> data model -> failure
handling -> trade-offs -> submit. Each is its own field, saved as it is typed. The
grader still reads one answer: the sections are joined under their headings, so the
rubric sees what the learner wrote in the shape they wrote it.
"""
from typing import Dict, Optional

SECTIONS = [
    ("requirements", "Requirements & scale assumptions"),
    ("architecture", "High-level architecture"),
    ("data_model", "Data model & storage"),
    ("failure_handling", "Failure handling"),
    ("trade_offs", "Trade-offs"),
]
SECTION_KEYS = [key for key, _ in SECTIONS]
SECTION_LABELS = dict(SECTIONS)


def clean_sections(raw: Optional[Dict[str, str]]) -> Optional[Dict[str, str]]:
    """Every known section, trimmed; None when none was given at all."""
    if raw is None:
        return None
    return {key: (raw.get(key) or "").strip() for key in SECTION_KEYS}


def compose_answer(sections: Dict[str, str]) -> str:
    """The sections under their headings, empty ones left out."""
    parts = [
        f"## {SECTION_LABELS[key]}\n{sections[key]}"
        for key in SECTION_KEYS
        if sections.get(key)
    ]
    return "\n\n".join(parts)

File: frontend/src/services/systemDesignSections.ts
// PrepBench - Copyright (c) 2026 Nimish Kanungo
// Licensed under the PolyForm Noncommercial License 1.0.0 (see LICENSE).
// Commercial use requires a separate licence from the copyright holder.

/**
 * The parts of a written system-design answer, in the order an interviewer
 * expects them. The keys match backend/app/services/system_design_sections.py,
 * which refuses any other.
 */

export type SectionKey = 'requirements' | 'architecture' | 'data_model' | 'failure_handling' | 'trade_offs';

export type Sections = Record<SectionKey, string>;

export const SECTIONS: { key: SectionKey; label: string; placeholder: string }[] = [
  {
    key: 'requirements',
    label: 'Requirements & scale assumptions',
    placeholder: 'Users, requests per second, read/write mix, payload sizes, peak load, latency and availability targets…',
  },
  {
    key: 'architecture',
    label: 'High-level architecture',
    placeholder: 'The components — gateway, queues or streams, workers, stores, third parties — and how a request moves through them…',
  },
  {
    key: 'data_model',
    label: 'Data model & storage',
    placeholder: 'What is stored, how it is keyed and partitioned, indexes, retention…',
  },
  {
    key: 'failure_handling',
    label: 'Failure handling',
    placeholder: 'Partitions, a downstream outage, retries and their limits, dead letters, backpressure…',
  },
  {
    key: 'trade_offs',
    label: 'Trade-offs',
    placeholder: 'What you chose, what you gave up for it, and when you would choose differently…',
  },
];

export const emptySections = (): Sections => ({
  requirements: '', architecture: '', data_model: '', failure_handling: '', trade_offs: '',
});

/** Sections from the server, with any missing key filled in empty. */
export const toSections = (raw?: Partial<Record<string, string>> | null): Sections => {
  const out = emptySections();
  if (raw) {
    for (const s of SECTIONS) out[s.key] = raw[s.key] ?? '';
  }
  return out;
};

export const hasContent = (sections: Sections): boolean =>
  SECTIONS.some((s) => sections[s.key].trim().length > 0);
